File: src/obstacle_avoidance.py
#!/usr/bin/env python

import rospy
import random
import argparse
from geometry_msgs.msg import Twist
from std_msgs.msg import Float32


def get_args():
    """ Get arguments from rosrun for individual deployment. """
    parser = argparse.ArgumentParser(
        description="Obstacle avoidance python script."
    )

    # Required arguments
    parser.add_argument("-n", "--number",
                        action="store",
                        required=False,
                        help="Add a pheeno number namespace.",
                        default="")

    # The rationale behind rospy.myargv()[1:] is provided here:
    # https://groups.google.com/a/rethinkrobotics.com/forum/#!topic/brr-users/ErXVWhRmtNA
    return parser.parse_args(rospy.myargv()[1:])


def random_turn():
    """ Random turn direction.

    Returns a random turn direction for the pheeno based on a uniform
    distribution.

    """
    if random.random() < 0.5:
        turn_direction = -0.05  # Turn left

    else:
        turn_direction = 0.05  # Turn right

    return turn_direction


def obstacle_check(msg, ir_location):
    global is_obstacle_in_way
    global sensor_triggered
    global sensor_limits
    if msg.data < sensor_limits[ir_location]:
        sensor_triggered = ir_location
        is_obstacle_in_way[ir_location] = True

    else:
        is_obstacle_in_way[ir_location] = False


if __name__ == "__main__":
    # Get arguments from argument parser.
    input_args = get_args()
    if input_args.number is "":
        pheeno_number = ""

    else:
        pheeno_number = "/pheeno_" + str(input_args.number)

    # Initialize Node
    rospy.init_node("pheeno_obstacle_avoidance")

    # Create Publishers and Subscribers
    pub = rospy.Publisher(
        pheeno_number + "/cmd_vel", Twist, queue_size=100)
    sub_ir_center = rospy.Subscriber(
        pheeno_number + "/scan_center", Float32, obstacle_check,
        callback_args="center")
    sub_ir_back = rospy.Subscriber(
        pheeno_number + "/scan_back", Float32, obstacle_check,
        callback_args="back")
    sub_ir_right = rospy.Subscriber(
        pheeno_number + "/scan_right", Float32, obstacle_check,
        callback_args="right")
    sub_ir_left = rospy.Subscriber(
        pheeno_number + "/scan_left", Float32, obstacle_check,
        callback_args="left")
    sub_ir_cr = rospy.Subscriber(
        pheeno_number + "/scan_cr", Float32, obstacle_check,
        callback_args="cr")
    sub_ir_cl = rospy.Subscriber(
        pheeno_number + "/scan_cl", Float32, obstacle_check,
        callback_args="cl")

    # Global Variables
    global is_obstacle_in_way
    global sensor_triggered
    global sensor_limits

    # Other important variables
    is_obstacle_in_way = {"center": False, "cr": False, "right": False,
                          "back": False, "left": False, "cl": False}
    sensor_triggered = 0
    sensors = {"center": 0.05, "cr": -0.05, "right": -0.05,
               "back": -0.05, "left": 0.05, "cl": 0.05}
    sensor_limits = {"center": 25.0, "cr": 10.0, "right": 10.0,
                     "back": 15.0, "left": 10.0, "cl": 10.0}
    cmd_vel_msg = Twist()
    obs_cmd_vel_msg = Twist()
    rate = rospy.Rate(2)
    count = 0

    while not rospy.is_shutdown():
        if True in is_obstacle_in_way.values():
            obs_cmd_vel_msg.linear.x = 0
            obs_cmd_vel_msg.angular.z = sensors[sensor_triggered]
            pub.publish(obs_cmd_vel_msg)

            # Prevent a random turn into an object after trying to avoid.
            count = 3

        else:
            if count is 0:
                cmd_vel_msg.linear.x = 0
                cmd_vel_msg.angular.z = random_turn()
                pub.publish(cmd_vel_msg)
                count += 1

            elif count is 3:
                cmd_vel_msg.linear.x = 0.05
                cmd_vel_msg.angular.z = 0
                pub.publish(cmd_vel_msg)
                count += 1

            elif count is 15:
                count = 0

            else:
                pub.publish(cmd_vel_msg)
                count += 1

        rate.sleep()

File: src/obstacle_avoidance.cpp
#include "ros/ros.h"
#include "std_msgs/Float32.h"
#include "geometry_msgs/Twist.h"
#include "pheeno_ros/command_line_parser.h"
#include "pheeno_ros/pheeno_robot.h"

int main(int argc, char **argv)
{
  // Initial Variables
  std::string pheeno_name;

  // Parse inputs
  CommandLineParser cml_parser(argc, argv);

  // Parse input arguments for Pheeno name.
  if (cml_parser["-n"])
  {
    std::string pheeno_number = cml_parser("-n");
    pheeno_name = "/pheeno_" + pheeno_number;
  }
  else
  {
    ROS_ERROR("Need to provide Pheeno number!");
  }

  // Initializing ROS node
  ros::init(argc, argv, "obstacle_avoidance_node");

  // Crate PheenoRobot Object
  PheenoRobot pheeno = PheenoRobot(pheeno_name);

  // ROS Rate loop
  ros::Rate loop_rate(15);

  // Variables before loop
  double linear = pheeno.getDefaultLinearVelocity();  // Initial vel is the default one provided by config file.
  double angular = 0.0;  // No turning.
  double max_range_to_avoid = 20.0;
  double min_range_to_avoid = 15.0;
  bool obs_flag = false;
  geometry_msgs::Twist cmd_vel_msg;

  while (ros::ok())
  {
    // Move and avoid obstacles if they are near.
    obs_flag = pheeno.avoidObstacleMove(linear, angular, max_range_to_avoid);
    obs_flag = pheeno.avoidObstacleStop(linear, angular, min_range_to_avoid);

    // Return values to default after obstacle avoided.
    if (!obs_flag)
    {
      linear = pheeno.getDefaultLinearVelocity();  // Default velocity
      angular = 0.0;
    }

    // Set values to cmd_vel msg
    cmd_vel_msg.linear.x = linear;
    cmd_vel_msg.angular.z = angular;

    // Publish, Spin, and Sleep
    pheeno.publishCmdVelocity(cmd_vel_msg);
    ros::spinOnce();
    loop_rate.sleep();
  }
}
